ibis/selectors.py

```diff
@@ -586,66 +606,113 @@ def if_all(selector: Selector, predicate: Deferred | Callable) -> IfAnyAll:
-        def pred(col: ir.Value) -> bool:
-            try:
-                (table,) = col.op().relations
-            except ValueError:
-                raise IbisError("Column should depend on exactly one table")
-
-            schema = table.schema
-            idxs = schema._name_locs
-            num_names = len(schema)
-            colname = col.get_name()
-            colidx = idxs[colname]
-
-            if isinstance(key, str):
-                return key == colname
-            elif isinstance(key, int):
-                return key % num_names == colidx
-            elif util.is_iterable(key):
-                return any(
-                    (isinstance(el, int) and el % num_names == colidx)
-                    or (isinstance(el, str) and el == colname)
-                    for el in key
-                )
-            else:
-                start = key.start or 0
-                stop = key.stop or num_names
-                step = key.step or 1
-
-                if isinstance(start, str):
-                    start = idxs[start]
-
-                if isinstance(stop, str):
-                    stop = idxs[stop] + 1
-
-                return colidx in range(start, stop, step)
-
-        return where(pred)
+    def __getitem__(self, key: str | int | slice | Iterable[int | str]):
+        if isinstance(key, slice):
+            key = Slice(key.start, key.stop, key.step)
+        return ColumnSlice(key)
 
 
 r = Sliceable()
 """Ranges of columns."""
 
 
+class First(Singleton, Selector):
+    def expand(self, table: ir.Table) -> Sequence[ir.Value]:
+        return [table[0]]
+
+    def expand_names(self, table: ir.Table) -> frozenset[str]:
+        return frozenset((table.columns[0],))
+
+
 @public
-def first() -> Predicate:
+def first() -> Selector:
     """Return the first column of a table."""
-    return r[0]
+    return First()
+
+
+class Last(Singleton, Selector):
+    def expand(self, table: ir.Table) -> Sequence[ir.Value]:
+        return [table[-1]]
+
+    def expand_names(self, table: ir.Table) -> frozenset[str]:
+        return frozenset((table.columns[-1],))
 
 
 @public
-def last() -> Predicate:
+def last() -> Selector:
     """Return the last column of a table."""
-    return r[-1]
+    return Last()
+
+
+class AllColumns(Singleton, Selector):
+    def expand(self, table: ir.Table) -> Sequence[ir.Value]:
+        return list(map(table.__getitem__, table.columns))
+
+    def expand_names(self, table: ir.Table) -> frozenset[str]:
+        return frozenset(table.columns)
 
 
 @public
-def all() -> Predicate:
+def all() -> Selector:
     """Return every column from a table."""
-    return r[:]
+    return AllColumns()
+
+
+class NoColumns(Singleton, Selector):
+    def expand(self, table: ir.Table) -> Sequence[ir.Value]:
+        return []
+
+    def expand_names(self, table: ir.Table) -> frozenset[str]:
+        return frozenset()
+
+
+@public
+def none() -> Selector:
+    """Return no columns."""
+    return NoColumns()
```

Comment:
When does this selector come up? Why might a user want to use it?

Reply:
In `pivot_wider`, check out #9912.

Reply:
Still not seeing why it's necessary. In that PR you're using it in a test, but wouldn't `id_cols=[]` work just as well? It's harmless, so happy to add it, but I still see when a user would reach for this.

Reply:
It's conceptually simpler, for both end users *and* developers to convert `id_cols` to a selector.

What selector would you convert `id_cols=[]` into?

If you're thinking that it'd be special-cased, well, then I guess I disagree with handling it that way, if we have an existing unifying concept that address it and we don't have to remember "oh yeah, and empty sequences are handled in this special way".

Reply:
I'm not actually handling the case of an empty sequence in `_to_selector`, so thanks for bringing this up!

Reply:
I think there's happy medium here: users who want to use `id_cols=[]` totally can, and the implementation of that as a selector is `none()`.

Reply:
> What selector would you convert id_cols=[] into?

Since everywhere we accept both selectors and lists of columns/expressions they have identical meaning, I'd expect `id_cols=[]` to be the same as `ibis.selectors.c()`. It looks like `_to_selector()` currently fails on `_to_selector([])` (which seems like a bug).

> Developers can work with a single kind of thing, and end users can reason about their input as a selector, or something that is convertible to one.

Right, I'd agree with that. But for the end user `[]` seems like a more natural way to say "no columns". Since it's coercible to a selector, any selector API should combine it just fine IMO and it feels more natural than typing `s.none()`.

Reply:
Thanks also for bringing up the `c()`-with-no-arguments-case, I'll add a test for that too.

Reply:
Ok, done!